File: scripts/ruflo-daemon-gc.py
#!/usr/bin/env python3
"""ruflo-daemon-gc — list and reap leaked ruflo/claude-flow daemons.

Pattern ported from pacphi/agentic-kit's daemons.mjs (MIT, Chris Phillipson):
registry-first discovery with a `ps` sweep fallback, staleness = workspace
gone OR older than the TTL, and a PID-REUSE GUARD — a PID is only signalled
after re-probing its live cmdline for `daemon start`; unconfirmable PIDs are
refused, never killed.

Context (agentbox): no ruflo daemon runs under supervisord and the runtime env
pins RUFLO_DAEMON_AI_WORKERS=0, so ANY daemon this finds was started ad hoc
inside a session (upstream ruflo #2661 background: leaked per-project daemons
once produced ~8.1B tokens/week). Upstream ruflo also self-reaps after a 12h
TTL — this tool is the belt-and-braces sweep between those reaps.

Usage:
  ruflo-daemon-gc.py              # preview (never signals)
  ruflo-daemon-gc.py --kill       # SIGTERM confirmed-stale daemons
  ruflo-daemon-gc.py --ttl 43200  # staleness age in seconds (default 12h)
  ruflo-daemon-gc.py --json       # machine-readable
"""

import argparse
import json
import os
import signal
import subprocess
import sys
import time

REGISTRY_FILES = ("ai-jobs.json", "workspace-leases.json", "repo-supervisors.json")


def proc_cmdline(pid):
    try:
        with open(f"/proc/{pid}/cmdline", "rb") as fh:
            return fh.read().replace(b"\0", b" ").decode("utf-8", "ignore").strip()
    except OSError:
        return None


def proc_age_seconds(pid):
    try:
        st = os.stat(f"/proc/{pid}")
        return time.time() - st.st_mtime
    except OSError:
        return None


def is_daemon_process(pid):
    """PID-reuse guard: only True when the LIVE cmdline is a ruflo daemon."""
    cmd = proc_cmdline(pid)
    if cmd is None:
        return None  # can't confirm — refuse to touch
    return "daemon start" in cmd and ("cli.js" in cmd or "ruflo" in cmd or "claude-flow" in cmd)


def registry_daemons(cf_home):
    found = {}
    for name in REGISTRY_FILES:
        path = os.path.join(cf_home, name)
        try:
            with open(path, encoding="utf-8") as fh:
                data = json.load(fh)
        except (OSError, json.JSONDecodeError):
            continue
        entries = data.values() if isinstance(data, dict) else data
        for e in entries if isinstance(entries, (list, type({}.values()))) else []:
            if not isinstance(e, dict):
                continue
            pid = e.get("pid")
            ws = e.get("workspace") or e.get("cwd") or e.get("repo") or "?"
            if isinstance(pid, int):
                found[pid] = {"pid": pid, "workspace": ws, "source": name}
    return found


def ps_sweep():
    found = {}
    try:
        out = subprocess.run(["ps", "axww", "-o", "pid=,args="],
                             capture_output=True, text=True, timeout=5).stdout
    except (OSError, subprocess.SubprocessError):
        return found
    for line in out.splitlines():
        parts = line.strip().split(None, 1)
        if len(parts) != 2 or "daemon start" not in parts[1]:
            continue
        if "cli.js" not in parts[1] and "ruflo" not in parts[1] and "claude-flow" not in parts[1]:
            continue
        try:
            pid = int(parts[0])
        except ValueError:
            continue
        ws = "?"
        if "--workspace " in parts[1]:
            ws = parts[1].split("--workspace ", 1)[1].split(" --")[0].strip()
        found[pid] = {"pid": pid, "workspace": ws, "source": "ps"}
    return found


def main():
    ap = argparse.ArgumentParser(description="List/reap leaked ruflo daemons (pid-reuse guarded).")
    ap.add_argument("--kill", action="store_true", help="SIGTERM confirmed-stale daemons")
    ap.add_argument("--ttl", type=int, default=43200, help="staleness age in seconds (default 43200 = 12h)")
    ap.add_argument("--json", action="store_true", help="machine-readable output")
    args = ap.parse_args()

    cf_home = os.path.expanduser("~/.claude-flow")
    daemons = registry_daemons(cf_home)
    for pid, d in ps_sweep().items():
        daemons.setdefault(pid, d)

    rows = []
    for pid, d in sorted(daemons.items()):
        confirmed = is_daemon_process(pid)
        if confirmed is False:
            continue  # registry entry whose PID now belongs to something else
        age = proc_age_seconds(pid)
        ws_gone = d["workspace"] not in ("?", "") and not os.path.isdir(d["workspace"])
        stale = ws_gone or (age is not None and age > args.ttl)
        rows.append({**d, "age_s": int(age) if age is not None else None,
                     "confirmed": bool(confirmed), "workspace_gone": ws_gone, "stale": stale})

    killed = []
    if args.kill:
        for r in rows:
            if not r["stale"]:
                continue
            if not r["confirmed"]:
                print(f"  refuse pid={r['pid']}: cmdline unconfirmable (pid reuse guard)", file=sys.stderr)
                continue
            if is_daemon_process(r["pid"]):  # re-probe immediately before signalling
                try:
                    os.kill(r["pid"], signal.SIGTERM)
                    killed.append(r["pid"])
                except OSError as e:
                    print(f"  pid={r['pid']}: kill failed ({e})", file=sys.stderr)

    if args.json:
        print(json.dumps({"daemons": rows, "killed": killed}, indent=2))
        return 0

    if not rows:
        print("✓ no ruflo/claude-flow daemons running")
        return 0
    for r in rows:
        mark = "STALE" if r["stale"] else "live "
        age = f"{r['age_s']}s" if r["age_s"] is not None else "?"
        gone = " (workspace gone)" if r["workspace_gone"] else ""
        print(f"  [{mark}] pid={r['pid']:>7} age={age:>9} src={r['source']:<22} {r['workspace']}{gone}")
    if killed:
        print(f"\nSIGTERM sent to {len(killed)} stale daemon(s): {killed}")
    elif any(r["stale"] for r in rows) and not args.kill:
        print("\nStale daemons found — re-run with --kill to reap them.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: services/agentbox-ops/src/bin/ruflo-daemon-gc.rs
//! `ruflo-daemon-gc` — list and reap leaked ruflo/claude-flow daemons.
//!
//! Replaces `scripts/ruflo-daemon-gc.py`. Registry-first discovery with a
//! process-table sweep fallback, staleness = workspace gone OR older than the
//! TTL, and a PID-REUSE GUARD: a PID is signalled only after re-probing its
//! live command line; unconfirmable PIDs are refused, never killed.
//!
//! No ruflo daemon runs under supervisord and the runtime pins
//! `RUFLO_DAEMON_AI_WORKERS=0`, so anything found here was started ad hoc
//! inside a session.

use agentbox_ops::procs;
use agentbox_ops::pyjson;
use clap::Parser;
use serde_json::{json, Value};
use std::collections::BTreeMap;
use std::path::Path;

const REGISTRY_FILES: [&str; 3] = [
    "ai-jobs.json",
    "workspace-leases.json",
    "repo-supervisors.json",
];

#[derive(Parser)]
#[command(
    name = "ruflo-daemon-gc",
    about = "List/reap leaked ruflo daemons (pid-reuse guarded)"
)]
struct Args {
    /// SIGTERM confirmed-stale daemons.
    #[arg(long)]
    kill: bool,
    /// Staleness age in seconds (default 43200 = 12h).
    #[arg(long, default_value_t = 43_200)]
    ttl: u64,
    /// Machine-readable output.
    #[arg(long)]
    json: bool,
}

/// A daemon-shaped process found in a registry file or the process table.
/// The PID is the map key, so it is not repeated here.
#[derive(Clone)]
struct Discovered {
    workspace: String,
    source: String,
}

/// Reads daemon PIDs recorded in the claude-flow registry files.
fn registry_daemons(cf_home: &Path) -> BTreeMap<u32, Discovered> {
    let mut found = BTreeMap::new();
    for name in REGISTRY_FILES {
        let Ok(text) = std::fs::read_to_string(cf_home.join(name)) else {
            continue;
        };
        let Ok(data) = serde_json::from_str::<Value>(&text) else {
            continue;
        };
        let entries: Vec<&Value> = match &data {
            Value::Object(map) => map.values().collect(),
            Value::Array(list) => list.iter().collect(),
            _ => continue,
        };
        for e in entries {
            let Some(pid) = e.get("pid").and_then(Value::as_u64) else {
                continue;
            };
            let workspace = ["workspace", "cwd", "repo"]
                .iter()
                .find_map(|k| e.get(*k).and_then(Value::as_str))
                .unwrap_or("?")
                .to_string();
            found.insert(
                pid as u32,
                Discovered {
                    workspace,
                    source: name.to_string(),
                },
            );
        }
    }
    found
}

fn main() {
    let args = Args::parse();
    let home = std::env::var("HOME").unwrap_or_else(|_| "/home/devuser".into());
    let cf_home = Path::new(&home).join(".claude-flow");

    let mut daemons = registry_daemons(&cf_home);
    for d in procs::sweep() {
        daemons.entry(d.pid).or_insert(Discovered {
            workspace: d.workspace,
            source: "ps".to_string(),
        });
    }

    let mut rows: Vec<Value> = Vec::new();
    let mut stale_confirmed: Vec<u32> = Vec::new();
    let mut stale_unconfirmed: Vec<u32> = Vec::new();

    for (pid, d) in &daemons {
        let confirmed = match procs::confirm_daemon(*pid) {
            // A registry entry whose PID now belongs to something else.
            Some(false) => continue,
            Some(true) => true,
            None => false,
        };
        let age = procs::run_time_secs(*pid);
        let workspace_gone =
            d.workspace != "?" && !d.workspace.is_empty() && !Path::new(&d.workspace).is_dir();
        let stale = workspace_gone || age.map(|a| a > args.ttl).unwrap_or(false);

        if stale {
            if confirmed {
                stale_confirmed.push(*pid);
            } else {
                stale_unconfirmed.push(*pid);
            }
        }

        rows.push(json!({
            "pid": pid,
            "workspace": d.workspace,
            "source": d.source,
            "age_s": age,
            "confirmed": confirmed,
            "workspace_gone": workspace_gone,
            "stale": stale,
        }));
    }

    let mut killed: Vec<u32> = Vec::new();
    if args.kill {
        for pid in &stale_unconfirmed {
            eprintln!("  refuse pid={pid}: cmdline unconfirmable (pid reuse guard)");
        }
        for pid in &stale_confirmed {
            // Re-probe immediately before signalling — the guard, not a formality.
            if procs::confirm_daemon(*pid) != Some(true) {
                eprintln!("  refuse pid={pid}: cmdline unconfirmable (pid reuse guard)");
                continue;
            }
            let Some(raw) = rustix::process::Pid::from_raw(*pid as i32) else {
                continue;
            };
            match rustix::process::kill_process(raw, rustix::process::Signal::Term) {
                Ok(()) => killed.push(*pid),
                Err(e) => eprintln!("  pid={pid}: kill failed ({e})"),
            }
        }
    }

    if args.json {
        println!(
            "{}",
            pyjson::dumps_indent(&json!({"daemons": rows, "killed": killed}), 2)
        );
        return;
    }

    if rows.is_empty() {
        println!("✓ no ruflo/claude-flow daemons running");
        return;
    }
    for r in &rows {
        let mark = if r["stale"].as_bool().unwrap_or(false) {
            "STALE"
        } else {
            "live "
        };
        let age = match r["age_s"].as_u64() {
            Some(a) => format!("{a}s"),
            None => "?".to_string(),
        };
        let gone = if r["workspace_gone"].as_bool().unwrap_or(false) {
            " (workspace gone)"
        } else {
            ""
        };
        println!(
            "  [{mark}] pid={:>7} age={age:>9} src={:<22} {}{gone}",
            r["pid"].as_u64().unwrap_or(0),
            r["source"].as_str().unwrap_or(""),
            r["workspace"].as_str().unwrap_or("")
        );
    }
    if !killed.is_empty() {
        println!(
            "\nSIGTERM sent to {} stale daemon(s): {killed:?}",
            killed.len()
        );
    } else if rows.iter().any(|r| r["stale"].as_bool().unwrap_or(false)) && !args.kill {
        println!("\nStale daemons found — re-run with --kill to reap them.");
    }
}
